File: ADT/dfs_bfs/2316.py
from collections import defaultdict


class Solution:
    def dfs(self, u, adj, visited):
        visited[u] = True
        sizeOfComponent = 1
        for v in adj[u]:
            if not visited[v]:
                sizeOfComponent += self.dfs(v, adj, visited)
        return sizeOfComponent

    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        adj = defaultdict(list)
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        visited = [False] * n
        remainingNodes = n
        result = 0
        for i in range(n):
            if not visited[i]:
                sizeOfComponent = self.dfs(i, adj, visited)
                result += sizeOfComponent * (remainingNodes - sizeOfComponent)
                remainingNodes -= sizeOfComponent
        return result

File: ADT/dfs_bfs/2316.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countPairs = function(n, edges) {
    const adj = Array.from({ length: n }, () => []);

    for (const [u, v] of edges) {
        adj[u].push(v);
        adj[v].push(u);
    }

    const visited = Array(n).fill(false);

    function dfs(u) {
        visited[u] = true;
        let size = 1;
        for (const v of adj[u]) {
            if (!visited[v]) {
                size += dfs(v);
            }
        }
        return size;
    }

    let remainingNodes = n;
    let result = 0;

    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            const size = dfs(i);
            result += size * (remainingNodes - size);
            remainingNodes -= size;
        }
    }

    return result;
};
